File: SnakeAndLadders.py
"""
Time Complexity : O(n ^ 2) where n is the length of the board
Space Complexity : O(n ^ 2) where n is the length of the board
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
from collections import deque
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        if len(board) == 0:
            return -1
        n = len(board)
        moves = [0]*(n*n)
        i = n - 1
        j = 0
        index = 0
        even = 0
        # Flattening out the board and decrementing the numerical values given as we
        # start our index from 0
        while i >= 0 and j >= 0:
            if board[i][j] == -1:
                moves[index] = board[i][j]
            else:
                moves[index] = board[i][j] - 1
            index += 1
            if even % 2 == 0:
                j += 1
                if j == n:
                    j -= 1
                    even += 1
                    i -= 1
            else:
                j -= 1
                if j < 0:
                    j += 1
                    even += 1
                    i -= 1
        q = deque([])
        moveCount = 0
        q.append(0)
        moves[0] = -2
        # Performing a BFS and recording all the visited places in the moves array 
        # itself for simplicity
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                if curr == n * n - 1: return moveCount
                for k in range(1, 7):
                    pos = curr + k
                    if pos > n * n - 1: break
                    if moves[pos] != -2:
                        if moves[pos] == -1:
                            moves[pos] = -2
                            q.append(pos)
                        else:
                            q.append(moves[pos])
                            moves[pos] = -2
            moveCount += 1
        return -1

File: SnakeAndLadders.java
//TC: O(n^2)
//SC: O(n^2)
class Solution {
    public int snakesAndLadders(int[][] board) {
        if(board == null || board.length == 0) return 0;
        int n = board.length;
        int[] arr = new int[n*n];
        int even = 0;
        int idx = 0; //on 1d arr
        int i = n-1;
        int j = 0;

        //fill 1D arr
        while(idx < arr.length){
            if(board[i][j] == -1){
                arr[idx] = board[i][j];
            }else{
                arr[idx] = board[i][j] - 1;
            }
            idx++;
            if(even % 2 == 0){
                //l to r
                j++;
                if(j == n){
                    i--;
                    j--;
                    even++;
                }
            }else{
                //r to l
                j--;
                if(j == -1){
                    i--;
                    j++;
                    even++;
                }
            }
        }

        //bfs
        Queue<Integer> q = new LinkedList<>();
        q.add(0); arr[0] = -2; //mark it visited
        int moves = 0;
        while(!q.isEmpty()){
            int size = q.size(); //require this bcz we increase number of moves after each level
            for(int l=0; l<size; l++){
                int curr = q.poll();
                if(curr == n*n-1) return moves;
                //roll the dice
                for(int k=1; k<7; k++){
                    int child = curr + k;
                    if(child < n*n){
                        if(arr[child] != -2){
                            if(arr[child] == -1){
                                q.add(child);
                            }else{
                                q.add(arr[child]);
                            }
                            arr[child] = -2; //mark it visited
                        }
                    }

                }
            }
            moves++;
        }
        return -1;
    }
}